lib/Doctrine/DBAL/TransactionManager.php

```diff
@@ -0,0 +1,275 @@
+<?php
+
+namespace Doctrine\DBAL;
+
+/**
+ * Manages transactions.
+ */
+class TransactionManager
+{
+    const SAVEPOINT_PREFIX = 'DOCTRINE2_SAVEPOINT_';
+
+    /**
+     * @var \Doctrine\DBAL\Connection
+     */
+    private $connection;
+
+    /**
+     * @var \Doctrine\DBAL\Transaction[]
+     */
+    private $activeTransactions = array();
+
+    /**
+     * Whether nested transactions should use savepoints.
+     *
+     * @var boolean
+     */
+    private $nestTransactionsWithSavepoints = false;
+
+    /**
+     * @param \Doctrine\DBAL\Connection $connection
+     */
+    public function __construct(Connection $connection)
+    {
+        $this->connection = $connection;
+    }
+
+    /**
+     * @return TransactionBuilder
+     */
+    public function createTransaction()
+    {
+        return new TransactionBuilder($this);
+    }
+
+    /**
+     * @param array $configuration
+     *
+     * @return \Doctrine\DBAL\Transaction
+     */
+    public function beginTransaction(array $configuration = array())
+    {
+        $this->connection->connect();
+
+        if (isset($configuration[TransactionBuilder::ISOLATION_LEVEL])) {
+            $this->connection->setTransactionIsolation($configuration[TransactionBuilder::ISOLATION_LEVEL]);
+        }
+
+        $transaction = new Transaction($this, $configuration);
+        $this->activeTransactions[] = $transaction;
```

Comment:
This might be nitpicking but what happens if beginning the transaction / creating the savepoint fails down below? The transaction manager remains in an inconsistent state. It still holds the transaction object even though it has not started.